### DISCUSSION TITLE
Updating the git revision of a single file

### DISCUSSION THREAD

**Discussion by felipecrs** [renovatebot/renovate#27705] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

I have [this dependency](https://github.com/felipecrs/jenkins-agent-dind/blob/9f657f8b201205567bafb32d90b86b1d8600d99e/scripts/prepare_image.sh#L134) in one of my projects, but from what I checked in Renovate's documentation what I want to achieve is not possible.

By the way this exact same situation also happens in [docker's own dockerfiles](https://github.com/docker-library/docker/blob/9ee39bafc3844108938e3469b262738a5b0e804b/25/dind/Dockerfile#L96).

```dockerfile
# https://github.com/docker/docker/tree/master/hack/dind
ENV DIND_COMMIT 65cfcc28ab37cb75e1560e4b4738719c07c6618e
```

I would like to only propose updates to this dependency with new git shas only when there is a change in the file, which can be tracked in this link: https://github.com/moby/moby/commits/master/hack/dind

![image](https://github.com/renovatebot/renovate/assets/29582865/0d40e152-6e22-4500-af22-bd1695435b4c)

Obviously the same can be done with git CLI is the repository is cloned.

Supposing Renovate does not support something like this today, perhaps it could be supported like:

```dockerfile
# renovate: datasource=github-file-history depName=moby/moby@master:hack/dind versioning=git
ENV DIND_COMMIT 65cfcc28ab37cb75e1560e4b4738719c07c6618e
```

Any thoughts?

### Logs (if relevant)

N/A

**Comment by rarkins** [comment]
What "algorithm" (using either `git` or REST API) would you use to determine that the SHA needs changing?

**Reply by felipecrs** [reply]
Perhaps something around basic support for `followTag`, where the tag would be the branch and would somehow need to be specified.

```console
❯ git log -n 1 --pretty=format:%H master -- hack/dind
65cfcc28ab37cb75e1560e4b4738719c07c6618e
```

```console
❯ curl -fsSL "https://api.github.com/repos/moby/moby/commits?sha=master&path=hack/dind&per_page=1" | jq -r '.[0].sha'
65cfcc28ab37cb75e1560e4b4738719c07c6618e
```

Nevertheless both methods would allow for sorting as well. You can get as many shas you wanted, sorted, through git log with the `-n` flag.

With the GitHub API, you can similarly control the with the `per_page` and `pages` query params. They also come sorted.

If the current SHA isn't latest, then it needs to be changed.

**Reply by rarkins** [reply]
This would maybe require a new datasource along the lines of "git-file" which has no versions but can return a digest. I don't think there's any way to support it today

**Reply by felipecrs** [reply]
Alright. Thank you for your insight. Maybe someday I'll try implementing it.
